Added time filter to query search page

vera-liu opened the pull request:
Based on feedback from the meeting, added from-to time filter for query search

<img width="1417" alt="screen shot 2016-10-12 at 2 11 09 pm" src="https://cloud.githubusercontent.com/assets/20978302/19327888/1ad1937c-9086-11e6-93f5-7e6e6c700cad.png">

vera-liu committed:
Added time filter to query search page

bkyryliuk commented:
@vera-liu  - could you also display query creation time in the search results ?

vera-liu committed:
Added start date

vera-liu commented:
@bkyryliuk Added

vera-liu committed:
Updated python endpoint test

vera-liu committed:
changed spec

bkyryliuk commented on caravel/assets/javascripts/SqlLab/common.js:
```diff
@@ -5,4 +5,27 @@ export const STATE_BSSTYLE_MAP = {
+  'running',
+];
+
+export const FROM_OPTIONS = [
```
could we reuse the options?

vera-liu replied:
Done:)

bkyryliuk commented on caravel/views.py:
```diff
@@ -2124,30 +2124,39 @@ def queries(self, last_updated_ms):
+        to_time = request.args.get('to')
 
-        if userId != 'null':
+        if user_id != 'null':
```
s/if user_id != 'null':/ if user_id

vera-liu replied:
Done:)

bkyryliuk commented on caravel/views.py:
```diff
@@ -2124,30 +2124,39 @@ def queries(self, last_updated_ms):
-        userId = request.args.get('userId')
-        databaseId = request.args.get('databaseId')
-        searchText = request.args.get('searchText')
+        user_id = request.args.get('userId')
```
s/user_id/search_user_id - not to confuse with the logged in user

vera-liu replied:
Done;)

bkyryliuk commented on caravel/views.py:
```diff
@@ -2124,30 +2124,39 @@ def queries(self, last_updated_ms):
+        if from_time != 'null':
+            query = query.filter(models.Query.start_time > int(from_time))
+
+        if to_time != 'null':
```
add unit test that show that to and from works

vera-liu replied:
Is there a way to generate fake queries to test this out?

bkyryliuk commented on caravel/views.py:
```diff
@@ -2124,30 +2124,39 @@ def queries(self, last_updated_ms):
-        databaseId = request.args.get('databaseId')
-        searchText = request.args.get('searchText')
+        user_id = request.args.get('userId')
+        database_id = request.args.get('databaseId')
```
s/.get('databaseId')/.get('database_id') - to be consistent
for now we use python conversion for args

vera-liu replied:
Done;)

vera-liu commented on caravel/assets/javascripts/SqlLab/components/QuerySearch.jsx:
```diff
@@ -34,13 +36,41 @@ class QuerySearch extends React.Component {
-  insertParams(baseUrl, params) {
-    return baseUrl + '?' + params.join('&');
+  getTimeFromSelection(selection) {
+    switch (selection) {
```
I think year might be tricky here since it's not guaranteed to be a fixed number of hours?

bkyryliuk replied:
what about storing the dict from option to the function that evaluates it ?

vera-liu replied:
I tried using the dict from value to function, but the function is evaluated when the dict is defined. There might be a way to delay the evaluation, but I figured switch statements might be a better option for mapping to functions?

bkyryliuk replied:
function should take the attr, not to be evaluated.
@mistercrunch @ascott  - what is your take here?

ascott replied:
i think switch/case works fine in this case. i think the function names could be made more clear. what is the advantage of using an object mapping over case/switch here @bkyryliuk ?

bkyryliuk replied:
@ascott  - less boilerplate.
```
getTimeFromSelection(selection) {
  return this.function_obj[selection]()
}
```
or smth like that

vera-liu commented on caravel/assets/spec/javascripts/sqllab/QuerySearch_spec.jsx:
```diff
@@ -15,7 +15,7 @@ describe('QuerySearch', () => {
   it('should have two Select', () => {
     const wrapper = shallow(<QuerySearch />);
-    expect(wrapper.find(Select)).to.have.length(2);
+    expect(wrapper.find(Select)).to.have.length(4);
```
hmm by state mutations do you mean tests for reducers? for now we haven't had much tests for functionalities, most are for rendering components in javascript

bkyryliuk replied:
more like tests that test the logic rather than presence of the components.
it you select option expect that state is updated, etc
if you execute search - except url is like expected

bkyryliuk commented on caravel/assets/javascripts/SqlLab/components/QuerySearch.jsx:
```diff
@@ -109,7 +141,27 @@ class QuerySearch extends React.Component {
+          <div className="col-sm-1">
+            <Select
+              name="select-from"
+              placeholder="[From-]"
```
internationalization

vera-liu replied:
could you give me an example?

vera-liu committed:
Added specs and tests

vera-liu commented:
Added more specs and tests as required. @bkyryliuk

bkyryliuk commented on caravel/assets/javascripts/SqlLab/components/QuerySearch.jsx:
```diff
@@ -63,10 +96,18 @@ class QuerySearch extends React.Component {
-      `searchText=${this.state.searchText}`,
-      `status=${this.state.status}`,
+      this.state.userId ?
+        `user_id=${this.state.userId}` : '',
```
it should fit into the same line

bkyryliuk commented on caravel/views.py:
```diff
@@ -2118,30 +2118,39 @@ def queries(self, last_updated_ms):
+        database_id = request.args.get('database_id')
+        search_text = request.args.get('search_text')
         status = request.args.get('status')
+        from_time = request.args.get('from')
```
please add a comment that time is the unix timestamp in seconds

bkyryliuk commented on caravel/views.py:
```diff
@@ -2118,30 +2118,39 @@ def queries(self, last_updated_ms):
+        if to_time:
+            query = query.filter(models.Query.start_time < int(to_time))
 
+        sql_queries = query.limit(config.get("QUERY_SEARCH_LIMIT")).all()
```
QUERY_SEARCH_LIMIT may be not set or absent in the config. Please handle that case.

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -373,13 +373,30 @@ def test_queries_endpoint(self):
+        resp = self.get_resp('/caravel/search_queries?'+'&'.join(params))
+        data = json.loads(resp)
+        self.assertEquals(1, len(data))
+        self.logout()
```
please test multiple use cases:
make sure that initially you have queries for 2 users, 1 user has queries for at least 2 databases

Test that:
- time filter filters out queries both for `to` and `from` use cases
- user_id filters given user queries
- QUERY_SEARCH_LIMIT works
- like query returns the expected queries

It would be nice to have all them as separate test case and you can you SetUp method to create initial dataset. Come by if you need some help with it.

ascott commented on caravel/assets/spec/javascripts/sqllab/DatabaseSelect_spec.jsx:
```diff
@@ -0,0 +1,25 @@
+import sinon from 'sinon';
+
+describe('DatabaseSelect', () => {
+  it('should render', () => {
```
could break this up into 2 tests: `it('is valid element')` and `it('renders')`

ascott commented on caravel/assets/spec/javascripts/sqllab/DatabaseSelect_spec.jsx:
```diff
@@ -0,0 +1,25 @@
+    ).to.equal(true);
+  });
+
+  it('should call onChange', () => {
```
could be more explicit with this test name...`it('calls onChange on select change')`

ascott commented on caravel/assets/spec/javascripts/sqllab/QuerySearch_spec.jsx:
```diff
@@ -5,26 +5,60 @@ import QuerySearch from '../../../javascripts/SqlLab/components/QuerySearch';
+    expect(wrapper.find(Select)).to.have.length(4);
+  });
+
+  it('should update userId', () => {
```
i generally think using `should` in tests is not neccessary. could make this more clear: `it('updates userId on user select change')`

ascott commented on caravel/assets/javascripts/SqlLab/common.js:
```diff
@@ -5,4 +5,18 @@ export const STATE_BSSTYLE_MAP = {
 };
 
-export const STATUS_OPTIONS = ['success', 'failed', 'running'];
+export const STATUS_OPTIONS = [
```
seems like this file should be called `constants` rather than `common` for clarity.

vera-liu committed:
Modified python/js tests and some function/file names
based on code review comments

vera-liu commented:
Addressed all comments in latest commit:
Done:
 - Addd tests for all filters for search endpoint
 - Modified specs
 -Changed function names for epoch time

@ascott @bkyryliuk

vera-liu committed:
Resolved conflicts in DashboardSelect_spec and QuerySearch_spec

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
+        data = json.loads(resp)
+        self.assertEquals(0, len(data))
+
+        self.login('gamma')
```
split it into separate functions

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
+        # Test search queries on status
+        resp = self.get_resp('/caravel/search_queries?status=success')
+        data = json.loads(resp)
+        self.assertEquals(2, len(data))
```
assert that data has queries with status == success

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
-        resp = self.client.get('/caravel/search_queries?'+'&'.join(params))
-        self.assertEquals(200, resp.status_code)
+        self.login('admin')
+        self.run_sql("SELECT * FROM ab_user", 'admin', client_id='client_id_1')
```
move query creation into the setUp function:
https://docs.python.org/2/library/unittest.html

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
+        params = [from_time, to_time]
+        resp = self.get_resp('/caravel/search_queries?'+'&'.join(params))
+        data = json.loads(resp)
+        self.assertEquals(1, len(data))
```
assert that data has a query that falls into given period of time

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
+        # Test search queries on text
+        resp = self.get_resp('/caravel/search_queries?search_text=permission')
+        data = json.loads(resp)
+        self.assertEquals(1, len(data))
```
assert that result query text has permission in it

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -354,13 +354,57 @@ def test_queries_endpoint(self):
+        self.assertEquals(1, len(data))
+        resp = self.get_resp('/caravel/search_queries?database_id=2')
+        data = json.loads(resp)
+        self.assertEquals(0, len(data))
```
assert that there are no queries in the db that have db id  = 2

vera-liu committed:
Break python tests for separate functions, Move sql queries to setUp()

vera-liu committed:
Get around eslint error for spec

vera-liu commented:
@bkyryliuk I updated the python tests, will move sqllab tests out of core_tests in a separate PR

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -41,6 +41,14 @@ def setUpClass(cls):
+        self.logout()
 
     def tearDown(self):
         pass
```
would be nice to drop queries you've created in the tear down

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
         self.login('admin')
         data = self.get_json_resp('/caravel/queries/{}'.format(0))
-        self.assertEquals(2, len(data))
+        self.assertEquals(4, len(data))
```
I see 5 queries created, 3 in setup and 2 here, why is the output 4?

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      data = json.loads(resp)
+      self.assertEquals(3, len(data))
+      for k, v in data.items():
+        assert data[k]['dbId'] == 1
```
s/assert/self.assertEquals

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      resp = self.get_resp('/caravel/search_queries?database_id=-1')
+      data = json.loads(resp)
+      self.assertEquals(0, len(data))
+      self.logout()
```
no need to logout

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      resp = self.get_resp('/caravel/search_queries?user_id={}'.format(user.id))
+      data = json.loads(resp)
+      self.assertEquals(1, len(data))
+      for (k, v) in data.items():
```
ditto, not need to loop through as there is just 1 query

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      resp = self.get_resp('/caravel/search_queries?status=success')
+      data = json.loads(resp)
+      self.assertEquals(2, len(data))
+      for k, v in data.items():
```
ditto

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      resp = self.get_resp('/caravel/search_queries?search_text=permission')
+      data = json.loads(resp)
+      self.assertEquals(1, len(data))
+      for k, v in data.items():
```
by the way if you don't use the var you can name it _
`for _, v in data.items()`

bkyryliuk commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      data = json.loads(resp)
+      self.assertEquals(2, len(data))
+      for k, v in data.items():
+          assert v['startDttm'] > int(first_query_time) \
```
use assertEquals,
when data = 2, you can actually check each item manually

vera-liu commented on tests/core_tests.py:
```diff
@@ -332,35 +340,101 @@ def test_queries_endpoint(self):
+      self.logout()
+
+    def test_search_query_on_user(self):
+      self.login('admin')
```
but we have access tests/other login logout in other tests, wouldn't they cause conflicts if we log in at Setup()?

vera-liu committed:
Small changes based on comments

vera-liu commented:
all done @bkyryliuk

bkyryliuk commented:
Lovely, thanks for bearing with me !
LGTM once the last comment is addressed.

vera-liu committed:
Merge branch 'master' into vliu_time_filter
